Add doc for TGI deployment (#557)
QwenLM/Qwen2: docs/source/deployment/tgi.rst

TGI
=====================

Hugging Face's Text Generation Inference (TGI) is a production-ready framework specifically designed for deploying and serving large language models (LLMs) for text generation tasks. It offers a seamless deployment experience, powered by a robust set of features:

> * **Speculative Decoding:**  Accelerates generation speeds.

Reviewer: If you advertise for those (speculative decoding and versatile device support, specifically), please at least provide a link to the documentation on how to use those.

* **Tensor Parallelism:**  Enables efficient deployment across multiple GPUs.
* **Token Streaming:**  Allows for the continuous generation of text.
* **Versatile Device Support:**  Works seamlessly with AMD, Gaudi, AWS Inferentia, and Google TPU.

Installation
-----------------

> The easiest way to use TGI is via the TGI docker image, but it's possible to run it locally via Conda. Please refer this doc for instructions on the installation: `https://github.com/huggingface/text-generation-inference?tab=readme-ov-file#local-install`

Reviewer 2: Maybe link to https://huggingface.co/docs/text-generation-inference/installation instead

Deploy Qwen with TGI
-----------------------

Find a model from qwen2 collection and get the model path: `https://huggingface.co/collections/Qwen/qwen2-6659360b33528ced941e557f`

You can either run the model via CLI or via Docker

> .. code:: bash

Reviewer 2: I would suggest to also include a Python example

Author: Not sure how to start TGI with python but added a Python client example.

   # CLI

>    text-generation-launcher --model-id Qwen/Qwen2-0.5B-Instruct --port 8080

Reviewer 2: This requires installation from source, so I would just showcase Docker here

   # via Docker
   model=Qwen/Qwen2-0.5B-Instruct
   volume=$PWD/data # share a volume with the Docker container to avoid downloading weights every run

   docker run --gpus all --shm-size 1g -p 8080:80 -v $volume:/data ghcr.io/huggingface/text-generation-inference:2.0 --model-id $model

Using TGI API
-------------

Once deployed, the model will be available on the selected port

.. code:: bash

   curl 127.0.0.1:8080/generate_stream \
           -X POST \
           -d '{"inputs":"What is Deep Learning?","parameters":{"max_new_tokens":20}}' \
           -H 'Content-Type: application/json'

It's also available on OpenAI style API.

.. code:: bash

   curl http://localhost:8080/v1/chat/completions -H "Content-Type: application/json" -d '{

>            "model": "Qwen/Qwen2-0.5B-Instruct",

Reviewer 2: I don't think the model needs to be specified (it can be an empty string or "tgi"I think)

           "messages": [

>            {"role": "system", "content": "You are a helpful assistant."},

Reviewer 2: nit: missing formatting

           {"role": "user", "content": "Tell me something about large language models."}
           ],
           }'

> The full API is listed in this Swagger UI: `https://huggingface.github.io/text-generation-inference/#/Text%20Generation%20Inference/completions`

Reviewer: that page has a huge error toast at the top

![image](https://github.com/QwenLM/Qwen2/assets/17811943/6a161a75-0e85-45a8-975b-6ffff197c281)

> Quantization for Performance

Reviewer: I would suggest that 7B is used consistently in this section.

Author: Sure

----------------------------

> Qwen2 come with both GPTQ and AWQ model. You can run them with additional --quantize flag

Reviewer 2: ```suggestion
Qwen2 comes with both GPTQ and AWQ models. You can run them with additional `--quantize` flag
```

. 

> TGI can also serve quantized qwen models, for example Qwen/Qwen2-0.5B-Instruct can be ran with "--quantize gptq"

Reviewer 2: ```suggestion
TGI can also serve quantized Qwen models, for example, Qwen/Qwen2-0.5B-Instruct can be ran with `--quantize gptq`
```

> TGI can also serve quantized Qwen models, for example, Qwen/Qwen2-0.5B-Instruct can be ran with `--quantize gptq`

Reviewer: Does it mean that TGI will quantize Qwen2-0.5B-Instruction online with `--quantize gptq` or a quantized model should be supplied? Please clarify.

.. code:: bash

   # CLI
   text-generation-launcher --model-id Qwen/Qwen2-72B-Instruct-GPTQ-Int4 --port 8080 --quantize gptq

   # via Docker
   model=Qwen/Qwen2-72B-Instruct-GPTQ-Int4
   volume=$PWD/data # share a volume with the Docker container to avoid downloading weights every run

   docker run --gpus all --shm-size 1g -p 8080:80 -v $volume:/data ghcr.io/huggingface/text-generation-inference:2.0 --model-id $model --quantize gptq

> If the model is quantized with awq, e.g. Qwen/Qwen2-7B-Instruct-AWQ please use --quantize awq

Reviewer 2: ```suggestion
If the model is quantized with AWQ, e.g. Qwen/Qwen2-7B-Instruct-AWQ please use `--quantize awq`
```

Multi accelerators Deployment
---------------------

> Please use --num-shard flag and set it value to the number of accelerators.

Reviewer 2: ```suggestion
Please use `--num-shard` flag and set its value to the number of accelerators.
```

> 

Reviewer: add example commands, please

Zero-Code Deployment with HF Inference Endpoints
---------------------------------------------------

For effortless deployment, leverage Hugging Face Inference Endpoints.

- **GUI interface** `https://huggingface.co/inference-endpoints/dedicated`
- **Coding interface:** `https://huggingface.co/blog/tgi-messages-api`

Once deployed, the endpoint can be used as usual.

Common Issue

> ----------------

Reviewer: I would also put --shm-size issue here.

Author: Explained in the Multi-accelerator session
